feat(provider/google): Support TCP Proxy Load Balancing

sayap opened the pull request:
Mostly copy-paste from SSL Proxy code.

sayap commented:
PR for clouddriver: https://github.com/spinnaker/clouddriver/pull/1727

danielpeach commented on app/scripts/modules/google/src/loadBalancer/configure/tcp/listener.html:
```diff
@@ -0,0 +1,28 @@
+      </div>
+    </div>
+  </ng-form>
+  <div class="form-group small" style="margin-top: 20px;">
```
This looks extraneous.

danielpeach commented on app/scripts/modules/google/src/loadBalancer/configure/tcp/listener.html:
```diff
@@ -0,0 +1,28 @@
+          <tr>
+            <td>
+              <input class="form-control input-sm"
+                     readonly
```
Should this be `readonly`?

sayap replied:
This should actually be a dropdown list, for both SSL and TCP Proxy. Will change

jtk54 commented:
Backend was merged: https://github.com/spinnaker/clouddriver/pull/1727. I think we should focus on getting this functionality in before worrying too much about the de-duplication between TCP and SSL LBs.

danielpeach commented on app/scripts/modules/google/src/loadBalancer/configure/tcp/gceCreateTcpLoadBalancer.controller.ts:
```diff
@@ -0,0 +1,220 @@
+  public stack: string;
+  public detail: string;
+  public loadBalancerName: string;
+  public portRange = '443';
```
Any reason not to match the default in cloud console (25)?

danielpeach commented on app/scripts/modules/google/src/loadBalancer/configure/ssl/listener.html:
```diff
@@ -23,9 +23,11 @@
-              <input class="form-control input-sm"
-                     readonly
-                     required ng-model="ctrl.loadBalancer.portRange"/>
+              <select class="form-control input-sm"
```
This doesn't seem to work properly - if regardless of the value I select here, the resulting frontend's port is 443. This is not a UI bug (I verified that the payload leaving Deck is sending the right port), but this should be resolved before merging this PR.

danielpeach commented on app/scripts/modules/google/src/serverGroup/configure/wizard/cloneServerGroup.gce.controller.js:
```diff
@@ -277,7 +279,16 @@ module.exports = angular.module('spinnaker.serverGroup.configure.gce.cloneServer
         .value();
 
-      return loadBalancerNames.concat(selectedSslLoadBalancerNames);
+      var selectedTcpLoadBalancerNames = _.chain(loadBalancerIndex)
```
s/var/let

danielpeach commented on app/scripts/modules/google/src/serverGroup/configure/wizard/loadBalancingPolicy/loadBalancingPolicySelector.component.ts:
```diff
@@ -41,19 +41,20 @@ class GceLoadBalancingPolicySelectorController implements ng.IComponentControlle
       const hasHttp = selected.find((loadBalancer: any) => get(index[loadBalancer], 'loadBalancerType') === 'HTTP');
 
-      if ((hasSsl && hasHttp)) {
+      if (((hasSsl || hasTcp) && hasHttp)) {
```
Please remove the extra set of parentheses here (the fault is in the original code, but no reason to duplicate it).

sayap commented:
Thanks for the review. Good catch on the issues 😅 

I have created PR for issue 1 and issue 3, but can't really reproduce issue 2 yet. I guess it could be related to issue 1?

danielpeach commented:
I consistently get an NPE here on `collectEntries`: https://github.com/spinnaker/clouddriver/blob/master/clouddriver-google/src/main/groovy/com/netflix/spinnaker/clouddriver/google/deploy/GCEUtil.groovy#L1275

Actually, I get this for SSL load balancers also. I'm guessing that's because I don't have any URL maps in my project right now.

@jtk54 can you look? I think we just need 
```
projectUrlMaps?.collectEntries { ... } ?: []
```

danielpeach commented:
Could you please center this field? I also don't think it should be a table - could just be a standalone input field.
![tcp-port](https://user-images.githubusercontent.com/13868700/28280761-c81b35aa-6af2-11e7-91a0-0420c539c07b.png)

Please also resolve the remaining comments in this PR, then we can get it merged. Thanks.

sayap commented:
Just pushed a commit that addresses the comments.

Regarding the request to center the port field, I make the field aligns with the others instead.

Regarding the port default of 443, I think it is more important to be consistent between SSL Proxy and TCP Proxy in spinnaker, compared to being consistent between cloud console and spinnaker. Also, our use case for TCP Proxy is to workaround the lack of mutual TLS support in the HTTPS Proxy, so port 443 works for me :)

jtk54 commented:
This PR depends on https://github.com/spinnaker/clouddriver/pull/1748 -- without it, creating a server group attached to TCP/SSL fails if not urlMaps exist. Looks like that bug has been dormant for a while. Otherwise this PR tests out fine for me.

jtk54 commented:
Dependent PR merged, this one needs a rebase then LGTM.

sayap commented:
Thanks, rebased.

jtk54 commented:
Thanks for your contribution.
